File: android/crypto/src/main/kotlin/com/kordar/ghostchat/core/crypto/MessagePadding.kt
package com.kordar.ghostchat.core.crypto

import java.security.SecureRandom
import java.util.Base64

object MessagePadding {

    /** Pad: 4-char length prefix + base64(message) + random padding to 256-byte boundary */
    fun pad(message: String, deterministicPadByte: Byte? = null): ByteArray {
        val b64 = Base64.getEncoder().encodeToString(message.toByteArray(Charsets.UTF_8))
        val lenPrefix = "%04d".format(b64.length)
        val content = (lenPrefix + b64).toByteArray(Charsets.UTF_8)
        val padTo = if (content.isEmpty()) 256 else ((content.size + 255) / 256) * 256
        val padLen = padTo - content.size

        val padding = if (deterministicPadByte != null) {
            ByteArray(padLen) { deterministicPadByte }
        } else {
            ByteArray(padLen).also { SecureRandom().nextBytes(it) }
        }
        return content + padding
    }

    /** Unpad: read 4-char length prefix, extract base64, decode */
    fun unpad(padded: ByteArray): String {
        if (padded.size < 4) return ""
        val lenStr = String(padded, 0, 4, Charsets.UTF_8)
        val b64Len = lenStr.toIntOrNull() ?: return ""
        if (padded.size < 4 + b64Len) return ""
        val b64Str = String(padded, 4, b64Len, Charsets.UTF_8)
        return String(Base64.getDecoder().decode(b64Str), Charsets.UTF_8)
    }
}

File: ios/Sources/GhostCrypto/MessagePadding.swift
import Foundation
import Security

public enum MessagePadding {

    /// Pad message: 4-char length prefix + base64(message) + random padding to 256-byte boundary.
    /// `deterministicPadByte`: if non-nil, use this byte for padding instead of random (test only).
    public static func pad(_ message: String, deterministicPadByte: UInt8? = nil) -> Data {
        let b64 = Data(message.utf8).base64EncodedString()
        let lenPrefix = String(format: "%04d", b64.count)
        let content = Data((lenPrefix + b64).utf8)
        let padTo = content.count == 0 ? 256 : ((content.count + 255) / 256) * 256
        let padLen = padTo - content.count

        var padding: Data
        if let byte = deterministicPadByte {
            padding = Data(repeating: byte, count: padLen)
        } else {
            padding = Data(count: padLen)
            padding.withUnsafeMutableBytes { buf in
                _ = SecRandomCopyBytes(kSecRandomDefault, padLen, buf.baseAddress!)
            }
        }
        return content + padding
    }

    /// Unpad: read 4-char length prefix, extract base64, decode.
    public static func unpad(_ padded: Data) -> String {
        guard padded.count >= 4 else { return "" }
        let lenStr = String(data: padded.prefix(4), encoding: .utf8) ?? "0"
        let b64Len = Int(lenStr) ?? 0
        guard padded.count >= 4 + b64Len else { return "" }
        let b64Str = String(data: padded[4..<(4 + b64Len)], encoding: .utf8) ?? ""
        guard let decoded = Data(base64Encoded: b64Str) else { return "" }
        return String(data: decoded, encoding: .utf8) ?? ""
    }
}
